coverage/html.py

```diff
@@ -32,6 +37,26 @@
 ]
 
 
+def serve_htmldir(htmldir):
+    """ Serve htmldirectory for coverage files"""
+    future = concurrent.futures.Future()
+
+
+    def start_server():
+        httpd = http.server.HTTPServer(('127.0.0.1', 0), http.server.SimpleHTTPRequestHandler)
+        future.set_result(httpd.server_port)
+        httpd.serve_forever()
+
+
+    thread = threading.Thread(target=start_server, daemon=True)
```

Comment:
I don't think you need a thread here, `webbrowser.open_new` will spawn a new process

Reply:
The webserver has to be started before `webbrowser.open_new()` is run, otherwise, it will timeout trying to connect to a port that isn't being listened on.

Reply:
I just tried this:

```python
import http.server
import webbrowser

def main():
    httpd = http.server.HTTPServer(('127.0.0.1', 0), http.server.SimpleHTTPRequestHandler)
    webbrowser.open_new(f'http://localhost:{httpd.server_port}/')
    httpd.serve_forever()

if __name__ == "__main__":
    main()
```

and it works because `http.server.HTTPServer` uses `socketserver.TCPServer` which synchronously binds the socket in the constructor: https://github.com/python/cpython/blob/3.8/Lib/socketserver.py#L445-L467

Reply:
neat, I thought I had tried that previously... hrm. looks like it is in python2.7 too.

Reply:
also the threading approach wouldn't fix the problem if HTTPServer did run with `bind_and_activate=False`, consider the following race:

| main                                                                   | Thread-1                                                                                                |
|------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------|
| `thread.start()`                                                       |                                                                                                         |
|                                                                        | `httpd = http.server.HTTPServer(('127.0.0.1', 0), http.server.SimpleHTTPRequestHandler)`                |
|                                                                        | `future.set_result(httpd.server_port)` |
| `webbrowser.open_new(f'http://localhost:{future.result()}/{htmldir}')` |                                                                                                         |
|                                                                        | `httpd.serve_forever()`                                                                                 |

Reply:
wouldn't it wait until the future result was set before it ran the webbrowser.open_new()?

Eitherway, i will take a look at fixing it, if interest is given to have it merged into coveragepy.

Reply:
@gtmanfred yes, but look at the table, it waits for the `future.set_result(` but not the `httpd.serve_forever()`

Reply:
ahh fair, thanks